### DISCUSSION TITLE
qtile base16 support

### DISCUSSION THREAD

**Discussion by aMOPel** [qtile/qtile#2773] [Ideas]
Hey, i dont know if you want to implement something rather specific like color themes for this 
self tinker wm, but i made an integration for base16 https://github.com/chriskempson/base16 color themes into qtile.

I thought maybe you can use it to make something more generic and better integrated out of it and possibly push that to the base16 repo.

my repo : https://github.com/aMOPel/base16-qtile-test

i really like qtile so far ;) thanks for that

**Comment by zordsdavini** [comment]
I have "theme" idea in my Qtile future list :) I thought about this like separate project with many themes to select. And integration into config maybe could be with less user manual work (all these **params['widget']['groupbox'])
I was thinking is it possible to change colors and other stuff like to find object, ex. widget/groupbox and assign theme. Then in user config it could be like few lines at the end 
```
import ThemeProvider from qtile.theme
 ThemeProvider('base-16')
```
Bang and you have everething in place.

But your work is awesome and as you can see - gives some thoughts

**Comment by aMOPel** [comment]
Base16 has like a 300 themes, wasnt sure if u knew that from what you wrote.

Anyway, they have a standard for themes which is about which color goes on which number from 1 to 16, originally for syntax highligthing.
Since it is not easy to have rather generic color themes make sense with the elements in qtile i thought you might have use at least for my linking from [color numbers like `base00` to qtile elements](https://github.com/aMOPel/base16-qtile-test/blob/master/colors.py), since i dont hardcode a specific theme into my config but only the base16 interface for colors.
And i believe if you want to use base16 you have to do this link anyway, with some thought spend on which color goes where.

Or maybe i didnt get quite where you were going with your idea :smile:

**Comment by zordsdavini** [comment]
I didn't know about Base16, but now I understand themes list in some other apps, ex. pywal :) Thanks

Maybe you didn't understand me correctly, I'll try explain more.
My idea was how colors can be applied to Qtile. You took the way by giving dictionary of all objects and then each object (ex. widget, layout) will use this dict but assignment should be done manually. 
I thought that maybe there is the way to assign without such manual work. Some wrapper to set colors to objects. Even with some config like layout configs take from one them and colors from other. Just brainstorming.

BTW. if we will have clear interface to do such things we could look for pywal integration.

My config use colors from Pywal and font from Xresources, so actually I use themes, too. You can look at my config https://github.com/zordsdavini/dotfiles/blob/master/.config/qtile/config.py .

**Comment by m-col** [comment]
@zordsdavini I think your idea would be along the right lines. I think it might make more sense if we pass our colors directly into a `libqtile.config.theme` object rather than the name of a provider. This way, users can import colors from any source in their config, then pass those to the themer. Anything else in qtile that has default colors for things can take their colors from this object. We can then override these colors if we want by passing an individual color directly to that object in the config, so it will use that one instead of whichever it would have taken from the themer object.

**Comment by aMOPel** [comment]
the thing is i only did it the way i did because i didnt want to go through the qtile code where the objects are initiated with their parameters from config.py. i rather used the interface of the config.py for that assignment but yes, the idea of this issue was to have something in the end that is just a few lines saying which theme should be used.
there is this way or the way of users applying colors manually in someway to the seperate color parameters in the objects.
for that to work though you need a coupling of the interface of the theme provider to the color parameters.

ex: pywal has 8 colors, base16 has 16 colors, some other thing has possibly some other amount of colors with what ever rule, which color is on which number. If for the user it is not meant to be necessary to apply all those colors in a way they like, there has to be some hardlink for each theme provider you support to the color parameters.
so for pywal its like color00 goes to bar.background
color01 goes to widget.sep.foreground
etc.
and for base16 its something different since the colors are ordered diffenrently and there are more

but this assignment of color parameter to color number has to be coded somewhere in a way that makes sense since you want you focus border to have high contrast to you background and stuff.

of course it should still be possible for the user to adjust the colors for seperate objects in their config.py, so a nice config object that gets imported into the rigth place would be nice, like
`# in config.py
keys = {...}
screens = {...}
colors = {
widget.TaskList.border_active = GREEN,
...}`

on the other hand if you dont want to provide a default assignment of colors for theme providers,
then all i can ask for is a better interface for colors and themes, like you said @m-col. I thought, though there wouldnt be too many of those big colorproviders, and giving some sane defaults for them is a nice thing for new users.
Sorry for writing so much :worried:

**Comment by aMOPel** [comment]
BTW way a thing i noticed when extracting all parameters for color:
you mix the use of 'color' and 'colour' in your parameter names and documentation

**Comment by zordsdavini** [comment]
this British/American war... I can't decide what side should I chouse :D

**Comment by m-col** [comment]
I guess the solution to mapping colors onto the right parameters might be the ability to specify colors to the themer by name. For example:
```python
from libqtile.config import themer
from mytheme import mycolors

themer.background = mycolors[0]
themer.foreground = mycolors[1] 
# and so on...
```
However I would argue in favor of sticking with a common, standardised mapping used by many programs: X resources colors 0-15. Then colors in any other format could be converted into these by the user from any source:
```python
themer.colors = mycolors  # if they are already in X resources format
# or
themer.colors[2] = mycolors[8]  # rearrange colors to match that expected by Qtile
themer.colors[3] = mycolors[9]

# don't like Qtile's mapping for a given item...
mybar = bar.Bar(..., background=mycolors[3])
```
This is what many other programs do.

**Comment by aMOPel** [comment]
@m-col sounds good to me, go for it then :b

**Comment by scratchmex** [comment]
> I guess the solution to mapping colors onto the right parameters might be the ability to specify colors to the themer by name. For example:
> 
> ```python
> from libqtile.config import themer
> from mytheme import mycolors
> 
> themer.background = mycolors[0]
> themer.foreground = mycolors[1] 
> # and so on...
> ```
> 
> However I would argue in favor of sticking with a common, standardised mapping used by many programs: X resources colors 0-15. Then colors in any other format could be converted into these by the user from any source:
> 
> ```python
> themer.colors = mycolors  # if they are already in X resources format
> # or
> themer.colors[2] = mycolors[8]  # rearrange colors to match that expected by Qtile
> themer.colors[3] = mycolors[9]
> 
> # don't like Qtile's mapping for a given item...
> mybar = bar.Bar(..., background=mycolors[3])
> ```
> 
> This is what many other programs do.

I'd like to work on this. Can I help you with something?

**Comment by tych0** [comment]
I think it's really just a matter of unifying the source of colors across the whole codebase. You can see widget_defaults for how we did it for widgets, but I agree having the same source of defaults for layouts and widgets would be useful.

**Comment by 20jam** [comment]
I'd love to see this feature added, I use base16 for everything and its amazing.

**Comment by m-col** [comment]
@scratchmex Are you still interested in this? If you are still interested (4 months later, sorry!) that would be a great PR. I think all it would need would be a new class in libqtile.config (Theme?) that we can use to store our colors.

What would be ideal is if it something that acts list-y but has a fixed and required length (maybe `collections.deque`? There might be something better), then we can just access colours at e.g. `theme[4]`. As for arguments, it could either accept none, or a full list of 16 colours, or a dict with items like `color3: #00ff00` as users might only want to set one color, rather than all 16.

We'd also need to update the config-reading code so that the qtile object will stick it at `qtile.theme`, or if not defined to take the default from `libqtile.config` and use that. I guess the next step after that would be to make things use it, like the widgets and layouts. It would be really cool if they used the colors from `qtile.theme` in a way that allowed users to change the colors on the fly though I'm not sure what the best approach for that might be.

**Comment by scratchmex** [comment]
I was thinking on this and to add to the discussion if using base16 or other schemas. We can make a general `config.theme` class where all the colors are stored and if you want you can override by setting them explicitely as `background=Theme.colors[3]`. The advantage of storing all the colors in a theme class is that you can write adapters which implementen different schemas, for example the base16 or pywal. Also, the adapters can implement default settings for common (if not all) of the widgets. The main difficulty with this is to somewhat "attach" the configuration to the whole qtile. Like the way django does it for its settings module.
What you do think? I can work on this this winter.

**Comment by tcld** [comment]
Good to see you're still around. I just recently discovered pywal and was somewhat interested in this topic, too. Haven't found the time yet to start anything, though. So I'll leave it to you.

Your approach sounds solid. I think it might be nice to have a few functions on the theme class like `load_xresources`, `load_json` (to directly read the pywal output, for example) or `reset_to_default`.
I guess the idea with a `config.Theme`-class is to have one central theme that you can modify on the fly, right? There may also be an approach where the user can swap out a theme for another instead of modifying the existing theme. In that case an abstract base class might make sense.

Please keep us updated, I am very much looking forward to this. :)

**Comment by tcld** [comment]
Since I am still experimenting with pywal and thus have extended my config-file quite a bit to get some rudimentary theming, I felt the urge to invest a little time into this. So I drew up a basic `Theme`-class [here](https://github.com/qtile/qtile/compare/master...tcld:feature_theming). Feel free to improve or ignore it. :) 

While doing this I noticed that some of the widgets seem to redefine colors when they probably shouldn't, so that stuff should maybe be cleaned up a little (I added comments for the colors in question in that huge dictionary).

I decided against forcing the user to use exactly 16 colors - the default will likely need much less, and if somebody wants to provide more colors, I don't see a reason to stop them.

So far I don't have a clear idea about the "distribution" of the colors into widgets etc. I guess that would be next, however.
